src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java

```diff
@@ -47,6 +51,8 @@
 
   private AtomicBoolean hasTriedToReadAuthenticationFromStorage = new AtomicBoolean(false);
 
+  private Exercise submittedForGrading;
```

Comment:
Since this gets set in one thread (the one in `SubmissionStatusUpdater`) and read in another (when an update occurs and `updateExercises` is called), it should at least be marked as `volatile` to make sure that the modification made by one thread becomes visible in other threads. However in this case, instead of volatile, some synchronization should be used, because the following race condition exists:

- User starts an update using the refresh button, which calls `updateExercises`.
- `SubmissionStatusUpdater` simultaneously calls `setSubmittedFromGrading` from another thread.
- Now there is a race condition in the `submittedForGrading != null` check in `updateExercises`.

I'll add the synchronization in my PR since it's a small change :+1:

Reply:
I now realized that only storing one exercise also has some issues in the very unlikely scenario of submitting two different exercises one after another. So I think there is a way to kill two birds with one stone here: using a `ConcurrentHashMap` allows storing multiple exercises in grading and it also includes synchronization.